### ISSUE TITLE
Forward container ports

### ISSUE THREAD

**Issue by ingro** [dokku/dokku#1528]
Hello! I'm trying to expose more ports from my app container to the outside but I can't find a way to make it work. What I want to achieve is that my app is still available on the default address, **myapp.mydomain.com**, but an internal service (a websocket) could be contacted on **myapp.mydomain.com:5050**.

I did not found anything on the docs, so I tried to add those lines to the nginx conf:
```
server {
  listen 5050;
  server_name $NOSSL_SERVER_NAME;

  location / {
    proxy_pass $DOKKU_APP_LISTEN_IP:5050;
  }
}
```

But it seems that the `$DOKKU_APP_LISTEN_IP` variable is undefined at the time the configuration is compiled cause it outputs an empty string, and so the entire app fails.

I'm missing something? Is there a way to achieve this or it is not possible actually?

**Comment by rvalyi** [comment]
@ingro, I think you could do a plugin with 2 hooks:
one to pass an extra port to expose to the docker command with a hook called docker-args with:
```
#!/usr/bin/env bash
set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x

STDIN=$(cat)
output=" -p 5000 -p 5050"
echo "$STDIN$output"
```

and another hook named nginx-pre-reload to append the required extra nginx reverse proxy to that new port:
```
#!/bin/bash
set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
APP="$1"
DOKKU_APP_LISTEN_IP=$(< "$DOKKU_ROOT/$APP/IP.web.1")
NOSSL_SERVER_NAME=...
  nginx_extra=$(cat <<EOF
server {
  listen 5050;
  server_name $NOSSL_SERVER_NAME;
  location / {
    proxy_pass $DOKKU_APP_LISTEN_IP:5050;
  }
}

echo "$ngnx_extra}" >> "$DOKKU_ROOT/$APP/nginx.conf"
```

Well I hope you get the idea.

**Comment by ingro** [comment]
Thanks for the heads up @rvalyi , I'm not very familiar in writing Dokku plugins but I will give it a try as soon as I get some time.

**Comment by josegonzalez** [comment]
@ingro did that work?

**Comment by ingro** [comment]
Didn't find time to test that approach, will report back as soon as I can!

**Comment by josegonzalez** [comment]
*bump*

**Comment by michaelshobbs** [comment]
closing due to lack of response. feel free to re-open

**Comment by abl** [comment]
I actually just hit the original issue.

DOKKU_APP_LISTEN_IP is not defined when the template is being built, but DOKKU_APP_LISTENER_IP is.

Searching the source, it looks like that's the only time DOKKU_APP_LISTENER_IP is used, so changing it to DOKKU_APP_LISTEN_IP would solve the problem.

**Comment by josegonzalez** [comment]
@abl can you provide more information about your setup?

**Comment by abl** [comment]
Sure - I had a different setup in that I wanted two different subdomains for two ports but I just changed it over to match the original issue.

Using InfluxDB as my example - specifically https://github.com/tutumcloud/influxdb/tree/master/0.9 - which has the advantage of being the thing I deployed :)

It exposes two ports; 8083 is the admin interface, 8086 is the API. 8083 gets picked up by Dokku correctly and bound to influxdb.example.com:443 so there's no work to do for it.

To expose :8086 as influxdb.example.com:8086 the following stanza needs to be added to nginx.conf.template (I did the change in /home/dokku/influxdb since I was lazy):

```
server {
  listen      [::]:8086 ssl spdy;
  listen      8086 ssl spdy;
  server_name $SSL_SERVER_NAME;
  access_log  /var/log/nginx/${APP}-access.log;
  error_log   /var/log/nginx/${APP}-error.log;
$SSL_DIRECTIVES

  keepalive_timeout   70;
  add_header          Alternate-Protocol  8086:npn-spdy/2;
  location    / {

    gzip on;
    gzip_min_length  1100;
    gzip_buffers  4 32k;
    gzip_types    text/css text/javascript text/xml text/plain text/x-component application/javascript application/x-javascript application/json application/xml  application/rss+xml font/truetype application/x-font-ttf font/opentype application/vnd.ms-fontobject image/svg+xml;
    gzip_vary on;
    gzip_comp_level  6;

    proxy_pass  http://influxapi;
    proxy_http_version 1.1;
    proxy_set_header Upgrade \$http_upgrade;
    proxy_set_header Connection "upgrade";
    proxy_set_header Host \$http_host;
    proxy_set_header X-Forwarded-Proto \$scheme;
    proxy_set_header X-Forwarded-For \$remote_addr;
    proxy_set_header X-Forwarded-Port \$server_port;
    proxy_set_header X-Request-Start \$msec;
  }
  include $DOKKU_ROOT/$APP/nginx.conf.d/*.conf;
}
upstream influxapi {
  server $DOKKU_APP_LISTENER_IP:8086;
}
```

Basically, copy the existing stanza and do the following:

1. Change `$NGINX_SSL_PORT` to `8086`
2. Create a new upstream entry based on $DOKKU_APP_LISTENER_IP
3. Change the `proxy_pass` directive to point at your new upstream.

**Comment by abl** [comment]
This has a few side effects, it looks like - for one it, it causes CHECKS to fail - possibly since you're trying to bind a port twice?

**Comment by michaelshobbs** [comment]
We don't currently have a way to persistently expose more than one port for a single deployed app. I think this is a major departure from our Heroku model and probably won't be addressed any time soon. @josegonzalez any additional thoughts here?

**Comment by josegonzalez** [comment]
It's something that comes up over and over, and is definitely a feature of other PaaS software. 

Fwiw, all the official plugins can expose multiple ports.

Any further discussion should take place in the nginx RFC.

**Comment by michaelshobbs** [comment]
Can you ref an example of exposing multi ports in the nginx rfc issue?
